// @polkadot/types/cjs/metadata/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Metadata", {
  enumerable: true,
  get: function () {
    return _Metadata.Metadata;
  }
});
Object.defineProperty(exports, "PortableRegistry", {
  enumerable: true,
  get: function () {
    return _PortableRegistry.PortableRegistry;
  }
});
Object.defineProperty(exports, "decorateConstants", {
  enumerable: true,
  get: function () {
    return _decorate.decorateConstants;
  }
});
Object.defineProperty(exports, "decorateExtrinsics", {
  enumerable: true,
  get: function () {
    return _decorate.decorateExtrinsics;
  }
});
Object.defineProperty(exports, "decorateStorage", {
  enumerable: true,
  get: function () {
    return _decorate.decorateStorage;
  }
});
Object.defineProperty(exports, "expandMetadata", {
  enumerable: true,
  get: function () {
    return _decorate.expandMetadata;
  }
});

var _Metadata = require("./Metadata");

var _PortableRegistry = require("./PortableRegistry");

var _decorate = require("./decorate");
